### PR TITLE
Add AnyHttpUrl validator for jenkins's server_url config field

### PR DESCRIPTION
**Pull request by Thanhphan1147**
Applicable spec: <link>

### Checklist

- [x] The [charm style guide](https://juju.is/docs/sdk/styleguide) was applied
- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The changes are compliant with [ISD054 - Managing Charm Complexity](https://discourse.charmhub.io/t/specification-isd014-managing-charm-complexity/11619)
- [x] The documentation is generated using `src-docs`
- [x] The documentation for charmhub is updated.
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### CODE REVIEW

File: src/state.py [hunk 1]
```diff
@@ -37,6 +37,13 @@ def __init__(self, msg: str = ""):
         """
         self.msg = msg
 
+class Validator(BaseModel):
+    """Pydantic validator for various attributes
+
+    Attrs:
+        http_url_validator: assign a value to this attribute to validate it against pydantic's HttpUrl type
```
Comment (gtrkiller) [inline src/state.py:44]: Shouldn't the type be AnyHttpUrl istead of HttpUrl in the docstring?

Comment (gtrkiller) [inline src/state.py:44]: Shouldn't the type be AnyHttpUrl istead of HttpUrl in the docstring?

File: src/state.py [hunk 2]
```diff
@@ -49,6 +56,19 @@ class JenkinsConfig(BaseModel):
     server_url: str = Field(..., min_length=1)
     agent_name_token_pairs: typing.List[typing.Tuple[str, str]] = Field(..., min_items=1)
 
+    @validator("server_url")
+    def valid_http_url(cls, server_url: str):
+        """Pydantic validator for sever_url attribute
```
Comment (gtrkiller) [inline src/state.py:61]: Small nitpick: sever_url -> server_url

**Review by gtrkiller — COMMENTED** [review]
LGTM! Some small comments

File: src/state.py [hunk 3]
```diff
@@ -49,6 +59,20 @@ class JenkinsConfig(BaseModel):
     server_url: str = Field(..., min_length=1)
```
[the remark this answers is not in the document]
Comment (Thanhphan1147) [reply src/state.py:60]: I think setting this to something other than str will change the dataclass, which will probably affect other parts of the charm that uses it. 

I can also add a property decorator and do some conversion to get the string, if you think that it's a better approach. maybe something like: 

```python
    @property
    def server_url(self) -> str:
        return self.url.__str__
```

[the remark this answers is not in the document]
Comment (Thanhphan1147) [reply src/state.py]: Changed it to the other method and it feel much cleaner!

**Review by arturo-seijas — COMMENTED** [review]
Can you add unit tests?

**Comment by amandahla** [comment]
Could you fill in the PR description?

**Review by yanksyoon — DISMISSED** [review]
Thx for the changes!

File: tests/unit/test_state.py [hunk 4]
```diff
@@ -71,6 +71,24 @@ def test_from_charm_invalid_charm_config(harness: ops.testing.Harness):
...
+):
+    """
+    arrange: given charm configuration data with invalid server_url attribute
+    act: when the state is initialized from_charm.
+    assert: InvalidStateError is raised. (due to pydantic's AnyHttpUrl validator)
+    """
+    invalid_config = config
+    # This configuration is invalid because schema (http or https) must be specified
+    invalid_config["jenkins_url"] = "example.com"
+    harness.update_config(invalid_config)
```
Comment (arturo-seijas) [inline tests/unit/test_state.py:85]: You won't be leveraging the harness so, can you mock the charm instead, as in
```
mock_charm = unittest.mock.MagicMock(spec=ops.CharmBase)
mock_charm.config = {"rjenkins_url": "example.com"}
```

Comment (arturo-seijas) [reply tests/unit/test_state.py:85]: Could you also fix the typo in the method name in line 92?

Comment (Thanhphan1147) [reply tests/unit/test_state.py:85]: I fixed the typo

Comment (arturo-seijas) [reply tests/unit/test_state.py:85]: Let's keep it as it is for now, then. But I'm not very keen on having the state relying on charm features too

**Review by arturo-seijas — DISMISSED** [review]
LGTM

**Review by arturo-seijas — APPROVED** [review]

**Review by yanksyoon — APPROVED** [review]
LGTM!

**Pull request merged**